src/thread.rs

```diff
@@ -321,37 +339,58 @@ impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
         }
     }
 
-    /// Allocate a new blockset id.
-    fn create_blockset(&mut self) -> BlockSetId {
-        self.blockset_counter = self.blockset_counter.checked_add(1).unwrap();
-        BlockSetId::new(self.blockset_counter)
-    }
-
-    /// Block the currently active thread and put it into the given blockset.
-    fn block_active_thread(&mut self, set: BlockSetId) {
-        let state = &mut self.active_thread_mut().state;
+    /// Put the thread into the blocked state.
+    fn block_thread(&mut self, thread: ThreadId) {
+        let state = &mut self.threads[thread].state;
         assert_eq!(*state, ThreadState::Enabled);
-        *state = ThreadState::Blocked(set);
+        *state = ThreadState::BlockedOnSync;
     }
 
-    /// Unblock any one thread from the given blockset if it contains at least
-    /// one. Return the id of the unblocked thread.
-    fn unblock_some_thread(&mut self, set: BlockSetId) -> Option<ThreadId> {
-        for (id, thread) in self.threads.iter_enumerated_mut() {
-            if thread.state == ThreadState::Blocked(set) {
-                trace!("unblocking {:?} in blockset {:?}", id, set);
-                thread.state = ThreadState::Enabled;
-                return Some(id);
-            }
-        }
-        None
+    /// Put the blocked thread into the enabled state.
+    fn unblock_thread(&mut self, thread: ThreadId) {
+        let state = &mut self.threads[thread].state;
+        assert_eq!(*state, ThreadState::BlockedOnSync);
+        *state = ThreadState::Enabled;
     }
 
     /// Change the active thread to some enabled thread.
     fn yield_active_thread(&mut self) {
         self.yield_active_thread = true;
     }
 
+    /// Register the given `callback` to be called once the `call_time` passes.
+    fn register_timeout_callback(
+        &mut self,
+        thread: ThreadId,
+        call_time: Instant,
+        callback: TimeoutCallback<'mir, 'tcx>,
+    ) {
+        self.timeout_callbacks
+            .insert(thread, TimeoutCallbackInfo { call_time: call_time, callback: callback })
+            .unwrap_none();
+    }
+
+    /// Unregister the callback for the `thread`.
+    fn unregister_timeout_callback_if_exists(&mut self, thread: ThreadId) {
+        self.timeout_callbacks.remove(&thread);
+    }
+
+    /// Get a callback that is ready to be called.
+    fn get_callback(&mut self) -> Option<(ThreadId, TimeoutCallback<'mir, 'tcx>)> {
+        let current_time = Instant::now();
+        // We use a for loop here to make the scheduler more deterministic.
```

Comment:
What does the loop have to do with determinism? If this about iteration order? If so, please say so.